File: android/src/main/java/com/ondato/OndatoConfiguration.kt
package com.ondato

import com.facebook.react.bridge.ReadableMap
import com.ondato.sdk.OndatoConfig
import com.ondato.sdk.enums.Language
import com.ondato.sdk.log.Level as OndatoLoggingLevel

// JS object → ReadableMap → OndatoConfiguration → OndatoConfig.Builder → Ondato.init(...)
data class OndatoConfiguration(
  val identityVerificationId: String,
  val mode: OndatoConfig.Mode,
  val language: Language?,
  val enableNetworkIssuesScreen: Boolean,
  val disablePdfFileUpload: Boolean,
  val switchPrimaryButtons: Boolean,
  val skipRegistrationIfDriverLicense: Boolean,
  val appearance: String?,
  val logLevel: OndatoLoggingLevel,
  val fonts: ReadableMap?,
  val requireScrollToEnableTermsButton: Boolean,
  val termsButtonTimeout: Long
) {
  companion object {
    fun fromReadableMap(map: ReadableMap): OndatoConfiguration {
      val id = map.getString("identityVerificationId") ?: ""
      val modeString = map.getString("mode") ?: "test"
      val mode =
        if (modeString.lowercase() == "live") OndatoConfig.Mode.LIVE else OndatoConfig.Mode.TEST

      val languageCode = map.getString("language")
      val language = when (languageCode?.lowercase()) {
        "bg" -> Language.Bulgarian
        "ca" -> Language.Catalan
        "cs" -> Language.Czech
        "de" -> Language.German
        "el" -> Language.Greek
        "en" -> Language.English
        "es" -> Language.Spanish
        "et" -> Language.Estonian
        "fi" -> Language.Finnish
        "fr" -> Language.French
        "hr" -> Language.Croatian
        "hu" -> Language.Hungarian
        "it" -> Language.Italian
        "lt" -> Language.Lithuanian
        "lv" -> Language.Latvian
        "nl" -> Language.Dutch
        "pl" -> Language.Polish
        "pt" -> Language.Portuguese
        "ro" -> Language.Romanian
        "ru" -> Language.Russian
        "sk" -> Language.Slovak
        "sl" -> Language.Slovenian
        "sq" -> Language.Albanian
        "sv" -> Language.Swedish
        "uk" -> Language.Ukrainian
        "vi" -> Language.Vietnamese
        else -> null
      }

      val logLevelString = map.getString("logLevel") ?: "info"
      val logLevel = when (logLevelString.lowercase()) {
        "error" -> OndatoLoggingLevel.Error
        "debug" -> OndatoLoggingLevel.Debug
        else -> OndatoLoggingLevel.Info
      }

      val androidFonts = if (map.hasKey("fonts")) {
        val fullFonts = map.getMap("fonts")
        if (fullFonts?.hasKey("android") == true) fullFonts.getMap("android") else null
      } else null

      return OndatoConfiguration(
        identityVerificationId = id,
        mode = mode,
        language = language,
        enableNetworkIssuesScreen = map.getBoolean("enableNetworkIssuesScreen"),
        disablePdfFileUpload = map.getBoolean("disablePdfFileUpload"),
        switchPrimaryButtons = map.getBoolean("switchPrimaryButtons"),
        skipRegistrationIfDriverLicense = map.getBoolean("skipRegistrationIfDriverLicense"),
        appearance = map.getString("appearance"),
        logLevel = logLevel,
        fonts = androidFonts,
        requireScrollToEnableTermsButton = map.getBoolean("requireScrollToEnableTermsButton"),
        termsButtonTimeout = map.getDouble("termsButtonTimeout").toLong()
      )
    }
  }

  fun toOndatoConfig(context: android.content.Context): OndatoConfig {
    val builder = OndatoConfig.Builder()
      .setIdentityVerificationId(identityVerificationId)
      .setMode(mode)
      .enableNetworkIssuesScreen(enableNetworkIssuesScreen)
      .disablePdfFileUploadForProofOfAddress(disablePdfFileUpload)
      .setSwitchPrimaryButtons(switchPrimaryButtons)
      .setSkipRegistrationIfDriverLicense(skipRegistrationIfDriverLicense)
      .setLoggingLevel(logLevel)
      .setTermsAndConditionsRules(requireScrollToEnableTermsButton, termsButtonTimeout)

    language?.let { builder.setLanguage(it) }

    appearance?.let { builder.setConfiguration(it) }

    fonts?.let { fonts ->
      val res = context.resources
      val pkg = context.packageName
      val getFontId = { key: String ->
        fonts.getString(key)?.let { name ->
          val id = res.getIdentifier(name, "font", pkg)
          if (id == 0) {
            android.util.Log.w(
              "OndatoModule",
              "Font resource '$name' not found for key '$key'"
            )
            null
          } else {
            id
          }
        }
      }

      val h1 = getFontId("title")
      val h2 = getFontId("subtitle")
      val body = getFontId("body")
      val list = getFontId("list")
      val label = getFontId("inputLabel")
      val btn = getFontId("button")

      // Only apply if at least one valid resource ID was resolved
      if (h1 != null || h2 != null || body != null || list != null || label != null || btn != null) {
        builder.setCustomFonts(
          heading1FontResource = h1,
          heading2FontResource = h2,
          normalFontResource = body,
          listFontResource = list,
          inputLabelFontResource = label,
          buttonTextFontResource = btn
        )
      }
    }

    return builder.build()
  }
}

File: ios/OndatoConfiguration.swift
import Foundation
import OndatoSDK
import React

struct OndatoConfiguration {
  let identityVerificationId: String
  let mode: OndatoEnvironment
  let language: OndatoSupportedLanguage
  let showTranslationKeys: Bool
  let skipRegistrationIfDriverLicense: Bool
  let showNoInternetConnectionView: Bool
  let disablePdfFileUpload: Bool
  let switchPrimaryButtons: Bool
  let appearance: String?
  let logLevel: Int

  static func fromDictionary(_ dict: NSDictionary) throws -> OndatoConfiguration {
    // Required fields
    guard let id = dict["identityVerificationId"] as? String, !id.isEmpty else {
      throw NSError(domain: "OndatoConfig", code: 0, userInfo: [NSLocalizedDescriptionKey: "identityVerificationId is required and must be non-empty"])
    }

    // Mode
    let mode: OndatoEnvironment = (dict["mode"] as? String)?.lowercased() == "live" ? .live : .test

    // Booleans with defaults
    let showTransKeys = dict["showTranslationKeys"] as? Bool ?? false
    let skipReg = dict["skipRegistrationIfDriverLicense"] as? Bool ?? false
    let showNoInternet = dict["enableNetworkIssuesScreen"] as? Bool ?? true
    let disablePdf = dict["disablePdfFileUpload"] as? Bool ?? false
    let switchButtons = dict["switchPrimaryButtons"] as? Bool ?? false

    // Language
    let languageCode = dict["language"] as? String
    let language: OndatoSupportedLanguage

    switch (languageCode) {
    case "bg": language = .BG
    case "ca": language = .CA
    case "cs": language = .CS
    case "de": language = .DE
    case "el": language = .EL
    case "en": language = .EN
    case "es": language = .ES
    case "et": language = .ET
    case "fi": language = .FI
    case "fr": language = .FR
    case "hu": language = .HU
    case "it": language = .IT
    case "lt": language = .LT
    case "lv": language = .LV
    case "nl": language = .NL
    case "pl": language = .PL
    case "pt": language = .PT
    case "ro": language = .RO
    case "ru": language = .RU
    case "sk": language = .SK
    case "sq": language = .SQ
    case "sv": language = .SV
    case "uk": language = .UA
    case "vi": language = .VI
    default:
      language = .EN
    }

    // Appearance JSON
    let appearance = dict["appearance"] as? String

    // Logging level
    let logLevelString = dict["logLevel"] as? String ?? "info"
    let logLevel: Int = {
      switch logLevelString.lowercased() {
      case "debug": return 1
      case "error": return 2
      default: return 0  // info
      }
    }()

    return OndatoConfiguration(
      identityVerificationId: id,
      mode: mode,
      language: language,
      showTranslationKeys: showTransKeys,
      skipRegistrationIfDriverLicense: skipReg,
      showNoInternetConnectionView: showNoInternet,
      disablePdfFileUpload: disablePdf,
      switchPrimaryButtons: switchButtons,
      appearance: appearance,
      logLevel: logLevel
    )
  }
}
